### PR TITLE
Update list view data flow

### PR DESCRIPTION
**Pull request by at-tienhoang**
#### Related issues (Required)
- Task #
- Bug #

#### Description (Required)
- [x] Create base diff
- [x] Change data flow

### CODE REVIEW

File: app/src/main/java/vn/asiantech/way/ui/base/Diff.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package vn.asiantech.way.ui.base
+
+import android.support.v7.util.DiffUtil
+
+class Diff<T>(private val oldList: List<T>, private val newList: List<T>) : DiffUtil.Callback() {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/base/Diff.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/base/Diff.kt#L5"></span>Public classes, interfaces and objects require documentation.</td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/ui/base/Diff.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        areItemsTheSameFunc = areItemsTheSame
+        return this
+    }
+
+    internal fun areContentsTheSame(areItemsTheSame: (oldItem: T, newItem: T) -> Boolean): Diff<T> {
+        areContentsTheSameFunc = areItemsTheSame
+        return this
+    }
+
+    internal fun calculateDiff(): DiffUtil.DiffResult {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/base/Diff.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/base/Diff.kt#L19"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/ui/base/Diff.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        return DiffUtil.calculateDiff(this)
+    }
+
+    override fun getOldListSize(): Int {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/base/Diff.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/base/Diff.kt#L23"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/ui/base/Diff.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        return oldList.size
+    }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/base/Diff.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/base/Diff.kt#L25"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/ui/base/Diff.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    override fun getNewListSize(): Int {
+        return newList.size
+    }
+
+    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/base/Diff.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/base/Diff.kt#L31"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/ui/base/Diff.kt [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        return areItemsTheSameFunc.invoke(oldList[oldItemPosition], newList[newItemPosition])
+    }
+
+    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/base/Diff.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/base/Diff.kt#L35"></span></td>
    </tr>
  </tbody>
</table>

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/base/Diff.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/base/Diff.kt#L35"></span>Checks whether files end with a line separator.</td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/ui/base/Diff.kt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        return areContentsTheSameFunc.invoke(oldList[oldItemPosition], newList[newItemPosition])
+    }
+
+
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/base/Diff.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/base/Diff.kt#L39"></span></td>
    </tr>
  </tbody>
</table>

File: app/src/main/java/vn/asiantech/way/ui/search/SearchActivity.kt [hunk 8]
```diff
@@ -3,6 +3,9 @@ package vn.asiantech.way.ui.search
 import android.app.ProgressDialog
 import android.content.Intent
 import android.os.Bundle
+import android.support.v7.util.DiffUtil.DiffResult
+import android.support.v7.util.ListUpdateCallback
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/search/SearchActivity.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/search/SearchActivity.kt#L7"></span>Unused import</td>
    </tr>
  </tbody>
</table>

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/search/SearchActivity.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/4937761f5a1f4726c7662f76441c6664a7db9349/app/src/main/java/vn/asiantech/way/ui/search/SearchActivity.kt#L7"></span>Unused import</td>
    </tr>
  </tbody>
</table>

**Comment by at-dev-support-01** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">All test passed. (17/17)</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: app/src/main/java/vn/asiantech/way/ui/search/SearchActivity.kt (hunk 8 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/search/SearchActivity.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/project-way/blob/7c744d41f78d05dad85878fc23eacfad5b3ac369/app/src/main/java/vn/asiantech/way/ui/search/SearchActivity.kt#L7"></span>Unused import</td>
    </tr>
  </tbody>
</table>

**Comment by at-tienhoang** [comment]
Do finally
Diff

**Pull request merged**